packages/widgets/src/blocks/legacy-widget/edit/preview.js

```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
  * WordPress dependencies
  */
 import { useRefEffect } from '@wordpress/compose';
-import { addQueryArgs } from '@wordpress/url';
-import { useState } from '@wordpress/element';
+import { useState, useEffect } from '@wordpress/element';
 import { Placeholder, Spinner, Disabled } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
 
 export default function Preview( { idBase, instance, isVisible } ) {
 	const [ isLoaded, setIsLoaded ] = useState( false );
+	const [ srcDoc, setSrcDoc ] = useState( '' );
+	const [ isPreviewFetched, setIsPreviewFetched ] = useState( false );
+
+	const abortController =
+		typeof window.AbortController === 'undefined'
+			? undefined
+			: new window.AbortController();
+
+	async function fetchPreviewHTML() {
+		const restRoute = `/wp/v2/widget-types/${ idBase }/render`;
+		return await apiFetch( {
+			path: restRoute,
+			method: 'POST',
+			signal: abortController?.signal,
+			data: {
+				id_base: idBase,
+				instance,
+			},
+		} );
+	}
+
+	useEffect( () => {
+		if ( ! isPreviewFetched ) {
+			setIsPreviewFetched( true );
+			fetchPreviewHTML()
+				.then( ( response ) => {
+					setSrcDoc( response.preview );
+				} )
+				.catch( ( error ) => {
+					if ( error.name === 'AbortError' ) {
+						// We don't want to log abort errors.
+						return;
+					}
+					window.console.error(
+						`An error occurred while trying to fetch preview: ${ error.message }`,
+						error
+					);
+				} );
+		}
+
+		return () => {
+			if ( typeof abortController !== 'undefined' ) {
+				abortController.abort();
+			}
+		};
+	}, [] );
 
 	// Resize the iframe on either the load event, or when the iframe becomes visible.
 	const ref = useRefEffect(
 		( iframe ) => {
 			// Only set height if the iframe is loaded,
 			// or it will grow to an unexpected large height in Safari if it's hidden initially.
-			if ( isLoaded ) {
-				// If the preview frame has another origin then this won't work.
-				// One possible solution is to add custom script to call `postMessage` in the preview frame.
-				// Or, better yet, we migrate away from iframe.
-				function setHeight() {
-					// Pick the maximum of these two values to account for margin collapsing.
-					const height = Math.max(
-						iframe.contentDocument.documentElement.offsetHeight,
-						iframe.contentDocument.body.offsetHeight
-					);
-					iframe.style.height = `${ height }px`;
-				}
+			if ( ! isLoaded ) {
```

Comment:
Should probably revert the change since that I think the diff is the same?

Reply:
The main idea behind this was to apply so called `return early` approach. We don't need to include all that code into parenthesis.
Please see example [here](https://github.com/jupeter/clean-code-php/blob/master/README.md#avoid-nesting-too-deeply-and-return-early-part-1). It's for PHP but it also applies to JavaScript.

Reply:
Yes, it definitely increases the diff size.
On the other hand I feel like creating a separate PR for it would be an overkill. We will have to create a separate issue for it as well.